src/ray/util/event.cc

```diff
@@ -16,11 +16,215 @@
 
 namespace ray {
 ///
+/// LogFileObject
+///
+LogFileObject::LogFileObject(std::string log_dir, std::string file_name)
+    : log_dir_(log_dir),
+      file_name_(file_name),
+      stop_writing_(false),
+      file_(NULL),
+      bytes_since_flush_(0),
+      file_length_(0),
+      rollover_attempt_(rollover_attempt_frequency_ - 1),
+      timer_(EventManager::Instance().GetIoService(),
+             boost::asio::chrono::seconds(loop_flush_wait_second_)) {
+  rotate_index_ = 0;
+  for (int i = 1; i <= max_rotate_num_; ++i) {
+    if (access(GetFileNameByIndex(i).c_str(), 0) != -1) {
+      rotate_index_++;
+    }
+  }
+  timer_.async_wait(boost::bind(&LogFileObject::LoopFlush, this));
+}
+
+LogFileObject::~LogFileObject() {
+  shutdown_ = true;
+  timer_.cancel();
+  absl::MutexLock lock(&mutex_);
+  if (file_ != NULL) {
+    fflush(file_);
+    fclose(file_);
+    file_ = NULL;
+  }
+}
+
+void LogFileObject::FlushUnlocked() {
+  if (file_ != NULL) {
+    fflush(file_);
+    bytes_since_flush_ = 0;
+  }
+}
+
+void LogFileObject::FlushLocked() {
+  absl::MutexLock lock(&mutex_);
+  FlushUnlocked();
+}
+
+void LogFileObject::LoopFlush() {
+  if (!shutdown_) {
+    FlushLocked();
+    if (!shutdown_) {
+      timer_.expires_at(timer_.expiry() +
+                        boost::asio::chrono::seconds(loop_flush_wait_second_));
+      timer_.async_wait(boost::bind(&LogFileObject::LoopFlush, this));
+    }
+  }
+}
+
+bool LogFileObject::CreateLogfile() {
+  file_ = fopen(GetFileNameByIndex().c_str(), "a+");
+  if (file_ == NULL) {
+    return false;
+  }
+  return true;
+}
+
+std::string LogFileObject::GetFileNameByIndex(int index) {
+  std::string result = log_dir_ + file_name_;
+  if (index == 0) {
+    result += ".log";
+    return result;
+  } else {
+    result += "." + std::to_string(index) + ".log";
+    return result;
+  }
+}
+
+void LogFileObject::Write(const char *message, int message_len, bool force_flush) {
+  absl::MutexLock lock(&mutex_);
+  if ((file_length_ >> 20) >= max_log_size_) {
+    if (file_ != NULL) {
+      fclose(file_);
+      // Rename file when the file size is excess max_log_size_
+      rotate_index_++;
```

Comment:
How to store rotate_index if worker crashed and then reconstructed again?
BTW,  so many low level api (FILE*/fclose/flush) have been used here. I'm not sure whether it can be transplanted to WINDOWS.

Reply:
First, if the worker reconstructed, log reporter will first detect all the log file in the event directory, then it will get the new rotate_index
```C++
rotate_index_ = 0;
  for (int i = 1; i <= max_rotate_num_; ++i) {
    if (access(GetFileNameByIndex(i).c_str(), 0) != -1) {
      rotate_index_++;
    }
  }
```

Second, it can not be transplanted to WINDOWS, I think. But if we just use the Glog, it‘s hard to define some complicated functions such as rotate, timing flush, and so on. Do you have some good idea(In Ant, We are going to use boost::log at the beginning, but there are compilation errors in some low versions of GCC)

Reply:
@Basasuya How to generate new rotate index? I mean you may not find the last rotate index if all suffix file names had been created. For example, max_rotate_num is 2, then both xx_1.log and xx_2.log will be in this directory.

For second issue, I test spdlog and it looks good for generating any format logging or something like that.

Reply:
I agree with your opinion, I have changed to the spdlog